### PR TITLE
Structuring the JSON form schema according to tenant.

### PR DESCRIPTION
**Pull request by BhagyashreeBhoyar**
Structuring form schema according to a tenant. Fetched form schema by using fetch API. Deleted member registration form schema json file and member registration correction form schema json file. Done a few refactoring.

### CODE REVIEW

File: src/components/routeComponents/MemberRegistrationForm.jsx [hunk 1]
```diff
@@ -33,6 +32,8 @@ import {
 import LinkButton from '../common/LinkButton';
 import Button from '../common/Button';
 import Form from '../form';
+import { fetchJsonSchemaFile } from '../../sagas/assetFilesAPI';
+import { getApplicationTenant } from '../../reducers/assetFilesReducer';
```
Comment (pepcus-pratik) [inline src/components/routeComponents/MemberRegistrationForm.jsx]: Again selector from incorrect reducer file

File: src/components/routeComponents/MemberRegistrationForm.jsx [hunk 2]
```diff
@@ -50,9 +51,33 @@ class MemberRegistrationForm extends Component {
...
+    try {
+      fetchJsonSchemaFile({ tenant, file: 'Registration' })
+        .then((response) => {
+          if (response) {
+            this.setState({
+              fileData: response,
+            });
+
+          } else {
+            this.props.setLoadingStateAction(false);
```
Comment (pepcus-pratik) [inline src/components/routeComponents/MemberRegistrationForm.jsx]: Use `finally` block

File: src/components/routeComponents/MemberRegistrationForm.jsx [hunk 3]
```diff
@@ -142,18 +166,17 @@ class MemberRegistrationForm extends Component {
   renderSuccessMessage = () => {
 
     const { isSubmitTriggered } = this.state;
-    const { isStudentCreated } = this.props;
+    const { isStudentCreated, newStudent } = this.props;
```
Comment (pepcus-pratik) [inline src/components/routeComponents/MemberRegistrationForm.jsx]: NO YJSG specific terms in the new code

File: src/components/routeComponents/MemberRegistrationForm.jsx [hunk 4]
```diff
@@ -142,18 +166,17 @@ class MemberRegistrationForm extends Component {
...
 
     if (isStudentCreated && isSubmitTriggered) {
-      const member = this.props.newStudent;
       // for pre-population on splash page
-      this.props.setStudentCredentialsAction(member.id, member.secretKey);
```
Comment (pepcus-pratik) [inline src/components/routeComponents/MemberRegistrationForm.jsx]: Why the **member** is renamed to **student** here?

File: src/components/routeComponents/MemberRegistrationForm.jsx [hunk 5]
```diff
@@ -142,18 +166,17 @@ class MemberRegistrationForm extends Component {
...
+      this.props.setStudentCredentialsAction(newStudent.id, newStudent.secretKey);
 
       return (
         <Popup>
           <p>{REGISTRATION_SUCCESS_MESSAGE}</p>
-          <p>{YOUR_ID_TEXT}<strong>{member.id}</strong>{IS_THERE_TEXT}</p>
-          <p>{YOUR_SECRET_CODE_TEXT}<strong>{member.secretKey}</strong>{IS_THERE_TEXT}</p>
+          <p>{YOUR_ID_TEXT}<strong>{newStudent.id}</strong>{IS_THERE_TEXT}</p>
```
Comment (pepcus-pratik) [inline src/components/routeComponents/MemberRegistrationForm.jsx]: Here again

File: src/components/routeComponents/MemberRegistrationForm.jsx [hunk 6]
```diff
@@ -183,37 +206,41 @@ class MemberRegistrationForm extends Component {
...
-            />
-          </div>
-        </Form>
-        {this.renderSuccessMessage()}
-      </div>
-    );
+    const { fileData } = this.state;
+    const { schema, UISchema, Data } = fileData;
+
+    if (!isEmpty(fileData)) {
```
Comment (pepcus-pratik) [inline src/components/routeComponents/MemberRegistrationForm.jsx]: Remove this condition.
The page should fail if the form configuration is not fetched successfully.

File: src/components/routeComponents/MemberRegistrationForm.jsx [hunk 7]
```diff
@@ -183,37 +206,41 @@ class MemberRegistrationForm extends Component {
...
-              value="Submit"
-              onClick={this.handleSubmit}
-            />
-          </div>
-        </Form>
-        {this.renderSuccessMessage()}
-      </div>
-    );
+    const { fileData } = this.state;
+    const { schema, UISchema, Data } = fileData;
```
Comment (pepcus-pratik) [inline src/components/routeComponents/MemberRegistrationForm.jsx]: Why did **UISchema** and **Data** start with an upper-case letter?

File: src/components/routeComponents/MemberRegistrationForm.jsx [hunk 8]
```diff
@@ -183,37 +206,41 @@ class MemberRegistrationForm extends Component {
...
+            formData={{ ...Data, ...this.state.member }}
+            onChange={this.onChange}
+            transformErrors={this.transformErrors}
+          >
+            <div className="register-button-wrapper">
+              {this.renderBackButton()}
+              <Button
+                buttonText={formSubmitBtnText}
+                type="submit"
+                formName=""
```
Comment (pepcus-pratik) [inline src/components/routeComponents/MemberRegistrationForm.jsx]: Why is this empty?

File: src/components/routeComponents/MemberRegistrationForm.jsx [hunk 9]
```diff
@@ -222,7 +249,9 @@ MemberRegistrationForm.propTypes = {
   createStudentDataAction: PropTypes.func,
   isStudentCreated: PropTypes.bool,
   newStudent: PropTypes.object,
+  setLoadingStateAction: PropTypes.string,
```
Comment (pepcus-pratik) [inline src/components/routeComponents/MemberRegistrationForm.jsx]: AGAIN. Are you copy-pasting the code?

File: src/components/routeComponents/MemberRegistrationForm.jsx [hunk 10]
```diff
@@ -231,13 +260,16 @@ MemberRegistrationForm.defaultProps = {
...
+  setLoadingStateAction: () => {},
   setStudentCredentialsAction: () => {},
+  tenant: '',
   userType: '',
 };
 
 const mapStateToProps = state => ({
   isStudentCreated: isCreated(state),
   newStudent: getNewStudent(state),
+  tenant: getApplicationTenant(state),
```
Comment (pepcus-pratik) [inline src/components/routeComponents/MemberRegistrationForm.jsx]:
```suggestion
  tenant: getTenantName(state),
```

File: src/config/tenant/BHOPAL/ADMIN.json [hunk 11]
```diff
@@ -0,0 +1,607 @@
+{
+  "Schema" : {
```
Comment (pepcus-pratik) [inline src/config/tenant/BHOPAL/ADMIN.json]:
```suggestion
  "schema" : {
```

File: src/config/tenant/BHOPAL/ADMIN.json [hunk 12]
```diff
@@ -0,0 +1,607 @@
...
+                  "509",
+                  "601",
+                  "603",
+                  "607"
+                ]
+              },
+              "address": {
+                "title": "पूरा पता",
+                "type": "string",
+                "validator": "addressValidation"
```
Comment (pepcus-pratik) [inline src/config/tenant/BHOPAL/ADMIN.json]: Let us set a practice to name all validator functions with the ***validator** suffix.

For Eg. addressValidator

File: src/config/tenant/BHOPAL/ADMIN.json [hunk 13]
```diff
@@ -0,0 +1,607 @@
...
+                  "Level 8 : छहढाला-ढाल 4-5"
+                ]
+              },
+              "marks2019": {
+                "title": "Marks 2019",
+                "type": "string"
+              },
+              "classRoomNo2019": {
+                "title": "Room",
+                "type": ["string"],
```
Comment (pepcus-pratik) [inline src/config/tenant/BHOPAL/ADMIN.json]: Why is in array?

File: src/config/tenant/BHOPAL/ADMIN.json [hunk 14]
```diff
@@ -0,0 +1,607 @@
...
+          "नहीं"
+        ],
+        "default": "Y"
+      },
+      "backButton": {
+        "type": "string"
+      },
+      "submitButton": {
+        "type": "string"
+      },
```
Comment (pepcus-pratik) [inline src/config/tenant/BHOPAL/ADMIN.json]: Do we need backButton and submitButton in the configuration?

@Nikmahajan

File: src/config/tenant/BHOPAL/ADMIN.json [hunk 15]
```diff
@@ -0,0 +1,607 @@
...
+        "title": "Marks 2017:",
+        "type": "string"
+      },
+      "marks2016": {
+        "title": "Marks 2016:",
+        "type": "string"
+      }
+    }
+  },
+  "UISchema" : {
```
Comment (pepcus-pratik) [inline src/config/tenant/BHOPAL/ADMIN.json]:
```suggestion
  "uiSchema" : {
```

File: src/config/tenant/BHOPAL/ADMIN.json [hunk 16]
```diff
@@ -0,0 +1,607 @@
...
+      "classNames": "field-style"
+    },
+    "email": {
+      "classNames": "field-style"
+    },
+    "classAttended2018": {
+      "classNames": "field-style"
+    },
+    "marks2019": {
+      "classNames": "field-style"
```
Comment (pepcus-pratik) [inline src/config/tenant/BHOPAL/ADMIN.json]: @Nikmahajan can we have a **fieldTemplate** for this to remove the redundant **field-style** code?

File: src/config/tenant/BHOPAL/ADMIN.json [hunk 17]
```diff
@@ -0,0 +1,607 @@
...
+    "classRoomNo2019": {
+      "classNames": "field-style",
+      "ui:placeholder": "कृपया एक चुनिए . . ."
+    },
+    "remark": {
+      "classNames": "field-style",
+      "ui:widget": "textarea"
+    }
+  },
+  "Data": {},
```
Comment (pepcus-pratik) [inline src/config/tenant/BHOPAL/ADMIN.json]: Is this required here??

File: src/config/tenant/BHOPAL/onlyEditOptIn.json [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+    {
+      "field": "email",
+      "validates": "optionalEmailValidation"
+    },
+    {
+      "field": "address",
+      "validates": "addressValidation"
+    }
+  ],
+  "defaultStudentDataFormat": [
```
Comment (pepcus-pratik) [inline src/config/tenant/BHOPAL/onlyEditOptIn.json]: In the future, we should remove this config.

Here is my suggestion,
- We'll have a JSON config for all tenants.
- We'll list out API endpoints in that JSON.
- We'll also have responseFormatter method defined in that JSON

This way all tenants have control over the data.

File: src/config/tenant/INDORE/Registration.json [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+      "mobile",
+      "busStop",
+      "address",
+      "classAttended2019"
+    ],
+    "properties": {
+      "name": {
+        "title": "नाम",
+        "type": "string",
+        "customValidator": "meraValidator"
```
Comment (pepcus-pratik) [inline src/config/tenant/INDORE/Registration.json]: Seriously!!!

File: src/config/tenant/INDORE/Registration.json [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+    },
+    "classAttended2019": {
+      "classNames": "field-style",
+      "ui:placeholder": "कृपया एक चुनिए . . ."
+    },
+    "classAttended2018": {
+      "classNames": "field-style"
+    }
+  },
+  "Data": {
```
Comment (pepcus-pratik) [inline src/config/tenant/INDORE/Registration.json]: Why do we need data in the JSON config??

@Nikmahajan

File: src/config/tenant/INDORE/Registration.json [hunk 21]
```diff
@@ -0,0 +1,262 @@
...
+    "optIn2019": "Y",
+    "classAttended2018": "",
+    "education": "",
+    "occupation": ""
+  },
+  "validation":
+  [
+    {
+      "field": "name",
+      "validates": "nameValidation"
```
Comment (pepcus-pratik) [inline src/config/tenant/INDORE/Registration.json]: `validates` is an incorrect name.

File: src/utils/formDataUtils.js [hunk 22]
```diff
@@ -28,109 +22,109 @@ export const getFormData = ({
...
-    } else if (pageUser === STUDENT || pageUser === STUDENT_WITH_URL) {
-      schema = indoreStudent.Schema;
+    } else if (isPageUserStudent({ pageUser })) {
+      schema = fileData.Schema;
       uiSchema = {
-        ...indoreStudent.UISchema,
+        ...fileData.UISchema,
         backButton: {
-          ...indoreStudent.UISchema.backButton,
+          ...fileData.UISchema.backButton,
```
Comment (pepcus-pratik) [inline src/utils/formDataUtils.js]: Rename all occurrences of **fileData** with a better alternative

File: src/components/routeComponents/MemberRegistrationCorrectionForm.jsx [hunk 23]
```diff
@@ -35,13 +34,15 @@ import {
   updateClassAttended2019InStudentData,
 } from '../../utils/registrationFormUtils';
 import { getApplicationTenant } from '../../reducers/assetFilesReducer';
-import { validation } from '../../config/memberRegistrationCorrectionFormSchema.json';
 import {
-  InitialStudentData, isObjectsEqual,
+  InitialStudentData,
+  isObjectsEqual,
   prePopulateOptIn,
 } from '../../utils/SampleFormValidation';
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: Done

File: src/components/routeComponents/MemberRegistrationCorrectionForm.jsx [hunk 24]
```diff
@@ -62,51 +63,136 @@ class MemberRegistrationCorrectionForm extends Component {
...
+  /**
+   * It fetch the JSON form schema for form
+   * @param {String} pageUser
+   * @param {String} tenant
+   * @param {String} file
+   * @param {Object} prePopulateOptInStudentData
+   * @param {Object} finalStudentData
+   * @param {Boolean} onlyOptInForm
+   */
+  fetchFileData = ({ pageUser, tenant, file, prePopulateOptInStudentData, finalStudentData, onlyOptInForm }) => {
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: Done.

[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationCorrectionForm.jsx [hunk 25]
```diff
@@ -62,51 +63,136 @@ class MemberRegistrationCorrectionForm extends Component {
...
+   * It decide that which JSON file will be fetch
+   * @param {Object} prePopulateOptInStudentData
+   * @param {Object} finalStudentData
+   * @param {Boolean} onlyOptInForm
+   * @return {Object}
+   */
+  DecideJsonFile = ({ prePopulateOptInStudentData, finalStudentData, onlyOptInForm }) => {
+    const { tenant, pageUser } = this.props;
+    const { ADMIN } = USER_TYPES;
+    if (isPageUserStudent({ pageUser }) && onlyOptInForm) {
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationCorrectionForm.jsx [hunk 26]
```diff
@@ -62,51 +63,136 @@ class MemberRegistrationCorrectionForm extends Component {
...
+   * @param {Boolean} onlyOptInForm
+   */
+  fetchFileData = ({ pageUser, tenant, file, prePopulateOptInStudentData, finalStudentData, onlyOptInForm }) => {
+    try {
+      fetchJsonSchemaFile({ tenant, file })
+        .then((response) => {
+          if (response) {
+            if (!isEmpty(prePopulateOptInStudentData)) {
+              this.setState({
+                student: InitialStudentData({
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: Done

File: src/components/routeComponents/MemberRegistrationCorrectionForm.jsx [hunk 27]
```diff
@@ -62,51 +63,136 @@ class MemberRegistrationCorrectionForm extends Component {
...
+              this.prePopulateCourse2019(InitialStudentData({
+                studentData: prePopulateOptInStudentData,
+                pageUser,
+                tenant,
+                onlyOptInForm,
+                fileData: response,
+              }));
+            }
+          } else {
+            this.props.setLoadingStateAction(false);
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationCorrectionForm.jsx [hunk 28]
```diff
@@ -62,51 +63,136 @@ class MemberRegistrationCorrectionForm extends Component {
...
-        oldStudentDate: InitialStudentData(finalStudentData),
-        isSubmitTriggered: false,
+  /**
+   * It decide that which JSON file will be fetch
+   * @param {Object} prePopulateOptInStudentData
+   * @param {Object} finalStudentData
+   * @param {Boolean} onlyOptInForm
+   * @return {Object}
+   */
+  DecideJsonFile = ({ prePopulateOptInStudentData, finalStudentData, onlyOptInForm }) => {
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: Done

File: src/components/routeComponents/MemberRegistrationCorrectionForm.jsx [hunk 29]
```diff
@@ -163,9 +249,11 @@ class MemberRegistrationCorrectionForm extends Component {
    * @param {String} value
    */
   changeIsOnlyOptIn = (value) => {
-    this.setState({
-      onlyOptInForm: value,
-    });
+    const finalStudentData = getFinalMemberData({ studentData: this.props.studentData });
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationCorrectionForm.jsx [hunk 30]
```diff
@@ -163,9 +249,11 @@ class MemberRegistrationCorrectionForm extends Component {
...
+    const prePopulateOptInStudentData = prePopulateOptIn(finalStudentData);
+
+    this.props.setLoadingStateAction(false);
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: No, I have removed it.

File: src/components/routeComponents/MemberRegistrationCorrectionForm.jsx [hunk 31]
```diff
@@ -62,51 +63,136 @@ class MemberRegistrationCorrectionForm extends Component {
       isSubmitTriggered: false,
       hasError: false,
       isFormChanged: false,
+      fileData: {},
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: Done

File: src/components/routeComponents/MemberRegistrationCorrectionForm.jsx [hunk 32]
```diff
@@ -355,13 +443,15 @@ class MemberRegistrationCorrectionForm extends Component {
           renderBackButton={this.renderBackButton}
           formRef={this.formRef}
           renderSubmitButtons={this.renderSubmitButtons}
+          fileData={fileData}
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: Done

File: src/components/routeComponents/MemberRegistrationCorrectionForm.jsx [hunk 33]
```diff
@@ -355,13 +443,15 @@ class MemberRegistrationCorrectionForm extends Component {
...
         >
           <FormUpdateSuccessMessage
             isSubmitTriggered={isSubmitTriggered}
             isFormChanged={isFormChanged}
             hasError={hasError}
             context={context}
             isStudentUpdated={isStudentUpdated}
+            isUpdatedReset={this.props.isUpdatedResetAction}
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: When member form was submitted successfully at that time isUpdatedReset method removed the login member data, id, and secret key.

File: src/components/routeComponents/MemberRegistrationCorrectionForm.jsx [hunk 34]
```diff
@@ -378,6 +468,7 @@ MemberRegistrationCorrectionForm.propTypes = {
   isUpdatedResetAction: PropTypes.func,
   pageUser: PropTypes.string,
   secretKey: PropTypes.string,
+  setLoadingStateAction: PropTypes.string,
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationCorrectionForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationForm.jsx [hunk 35]
```diff
@@ -33,6 +32,8 @@ import {
 import LinkButton from '../common/LinkButton';
 import Button from '../common/Button';
 import Form from '../form';
+import { fetchJsonSchemaFile } from '../../sagas/assetFilesAPI';
```
[the remark this answers is not in the document]
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationForm.jsx (hunk 1 above)
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationForm.jsx]: Done

File: src/components/routeComponents/MemberRegistrationForm.jsx (hunk 2 above)
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationForm.jsx (hunk 3 above)
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationForm.jsx (hunk 4 above)
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationForm.jsx (hunk 5 above)
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationForm.jsx (hunk 7 above)
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationForm.jsx (hunk 8 above)
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationForm.jsx (hunk 9 above)
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationForm.jsx]: Done.

File: src/components/routeComponents/MemberRegistrationForm.jsx (hunk 10 above)
Comment (BhagyashreeBhoyar) [reply src/components/routeComponents/MemberRegistrationForm.jsx]: Done.

File: src/config/tenant/BHOPAL/ADMIN.json (hunk 11 above)
Comment (BhagyashreeBhoyar) [reply src/config/tenant/BHOPAL/ADMIN.json]: Done.

File: src/config/tenant/BHOPAL/ADMIN.json (hunk 12 above)
Comment (BhagyashreeBhoyar) [reply src/config/tenant/BHOPAL/ADMIN.json]: Done.

File: src/config/tenant/BHOPAL/ADMIN.json (hunk 13 above)
Comment (BhagyashreeBhoyar) [reply src/config/tenant/BHOPAL/ADMIN.json]: Done.

File: src/config/tenant/BHOPAL/ADMIN.json (hunk 15 above)
Comment (BhagyashreeBhoyar) [reply src/config/tenant/BHOPAL/ADMIN.json]: Done.

File: src/config/tenant/INDORE/Registration.json (hunk 19 above)
Comment (BhagyashreeBhoyar) [reply src/config/tenant/INDORE/Registration.json]: Done.

File: src/config/tenant/INDORE/Registration.json (hunk 21 above)
Comment (BhagyashreeBhoyar) [reply src/config/tenant/INDORE/Registration.json]: Done.

File: src/utils/formDataUtils.js (hunk 22 above)
Comment (BhagyashreeBhoyar) [reply src/utils/formDataUtils.js]: Done.

**Pull request closed without merging**